keps/sig-etcd/4578-feature-gate/README.md

```diff
@@ -0,0 +1,437 @@
+
+We are proposing to add [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) in the similar way as Kubernetes does, so that can turn on or off any feature using the `--feature-gates` flag.
+
+Benefits from this enhancement includes:
+
+* graduating a feature is no longer a breaking change with new flags.
+
+* well-established feature lifecycle convention from Kubernetes.
+
+* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.
+
+* easy to query if a feature is available for an ectd cluster.
+
+### Goals
+
+* Introduce the `--feature-gates` flag in etcd code base.
+
+* Introduce apis/methods to query the enablement state of a feature.
+
+* Establish lifecycle stages of a feature, and clear criteria of lifecycle progression.
+
+### Non-Goals
+
+* Migrate all existing `--experimental` features to feature gate. 
+  * While feature gate provides a [path to deprecate existing `--experimental` features]((#path-to-migrate-existing-experimental-features)), and we will migrate several `--experimental` during the feature gate implementation as examples, the completion of the KEP does not require migrating all existing `--experimental` features. It can be done as followup tasks.
+
+* Use feature gate as a mechanism to fix bugs.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+It would be easier for developers to add new features without worrying too much about breaking etcd, because the feature would be Alpha first and disabled by default, and new changes would be mostly gated behind the feature gate.
+
+It would be a smoother process to move the feature to the next stage without the need to introduce a potentially breaking change of an `--experimental` flag.
+
+The developer also would not need to worry about feature compatibility when there is a mixed version cluster.
+
+#### Story 2
+
+Users of etcd cluster would be able to find out which features are enabled in the etcd cluster and decide on how to use them downstream. 
+
+Currently Kubernetes uses [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to check if a feature is supported by etcd, which basically checks the etcd version and compares the version with a hard-coded map of feature availability for different etcd versions.
+
+#### Story 3
+
+In a HA cluster, users should be able to enable/disable a feature without bring down the whole cluster, by restarting the servers with the new feature gate flag one by one. 
+
+#### Story 4
+
+During cluster upgrade/downgrade, feature changes across versions should have predictable behavior in this mix version scenario. 
```

Comment:
Can we assert a stronger claim/invariant here, say - `predictable behavior governed by the lowest member version`? Using concrete stories early on in the KEP can improve the mental model for what's proposed later on.

Reply:
Also if there's a good reason why we don't want to be that specific here, open to hearing it :)

Reply:
Cluster level features were dropped from the KEP scope, no longer need to consider upgrade/downgrade.